### PR TITLE
Feat(Dashboard): add multiple products in main dashboard

### LINKED ISSUE #218: Add products (main dashboard)
Description
Implement a backend feature to allow the creation of new multiple product at a time in the database. This endpoint will facilitate adding new products with details such as name, description, price, and category.

Acceptance Criteria
 A POST endpoint is available at /api/v1/products/add-products.
 The endpoint successfully creates list of new products and stores it in the database.
 Appropriate HTTP status codes are returned (e.g., 201 Created for success, 400 Bad Request for invalid input).
 Ensure validation of input data to prevent malformed or incomplete product entries.
Purpose
Provide an API endpoint for creating new products to facilitate the addition of products to the system.

Requirements
Create a backend endpoint to handle products creation requests.
Validate input data to ensure all necessary fields are provided and correctly formatted.
Ensure proper error handling and response statuses.
Update any related data or references accordingly.

Data validation
Input Validation:

Ensure all necessary fields (e.g., name, description, price, category) are provided and correctly formatted.

Validate that price is a positive number.

Output Validation:

Verify that the response includes the created product's details with correct data types.
Authentication and Authorization
Authentication:

Ensure the endpoint requires user authentication to create products.
Authorization:

Implement role-based access control to ensure only authorized users can create products.
Expected Outcome
A POST endpoint at /api/v1/products/add-products that:

Accepts product data (e.g., name, description, price, category, etc.).
Validates the input data.
Inserts the new product into the database.
Returns a 201 Created status with the created product details.
Returns appropriate error messages and statuses for invalid input or server errors.
Successful Response (201 Created):
{
 "status": "success",
 "message": "Products created successfully",
 "data": [{
   "id": 123,
   "name": "New Product",
   "description": "A description of the new product.",
   "price": 29.99,
   "category": "Electronics",
   "created_at": "2024-07-17",
   "updated_at": "2024-07-17"
 }]
}
Error Response (400 Bad Request):
{
  "status_code": 400,
  "status": "Bad request",
  "message": "Invalid input data",
  "errors": {
    "name": "Name is required",
    "price": "Price must be a positive number"
  }
}
Error Response (401 Unauthorized)
{
  "status_code": 401,
  "status": "unauthorized",
  "message": "Unauthorized. Please log in."
}
Error Response (500 Internal Server Error):
{
  "status_code": 500,
   "status": "Internal server error",
  "message": "An unexpected error occurred. Please try again later."
}
Testing Requirements
Unit Tests:

Verify the need for unit tests to cover all functionality of the endpoint.
Integration Tests:

Confirm the requirement for integration tests to ensure the endpoint works correctly with the overall system.

### PR DESCRIPTION
**Pull request by AlakaTahir**
# Description

Closes #218 

# Changes proposed

## What were you told to do?

I added endpoint to add multiple products

# Check List (Check all the applicable boxes)

🚨Please review the [contribution guideline](CONTRIBUTING.md) for this repository.

- [ ] My code follows the code style of this project.
- [ ] This PR does not contain plagiarized content.
- [ ] The title and description of the PR is clear and explains the approach.
- [ ] I am making a pull request against the **dev branch** (left side).
- [ ] My commit messages styles matches our requested structure.
- [ ] My code additions will fail neither code linting checks nor unit test.
- [ ] I am only making changes to files I was requested to.

# Screenshots/Videos

### CODE REVIEW

**Review by chimaihueze — CHANGES_REQUESTED** [review]
follow the PR naming standards, and write a detailed description for your PR.

File: src/Hng.Web/Controllers/ProductController.cs [hunk 1]
```diff
@@ -35,6 +35,21 @@ public async Task<ActionResult<ProductDto>> CreateProduct([FromBody] ProductCrea
             return Ok(successResponse);
         }
 
+        [HttpPost("add-products")]
+        [Authorize]
+        [ProducesResponseType(typeof(ProductDto), StatusCodes.Status201Created)]
+        public async Task<ActionResult<ProductsDto>> AddProducts([FromBody] AddMultipleProductDto body)
+        {
+            var loggedInUserId = HttpContext.User.FindFirst(ClaimTypes.Sid).Value;
```
Comment (chimaihueze) [inline src/Hng.Web/Controllers/ProductController.cs:43]: There is a service that already does this. Check the AuthenticationService in Infrastructure. You should move whatever logic you want to perform with this to the handler. Keep this controller simple.

File: src/Hng.Application/Features/Products/Commands/AddProductsCommand.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Hng.Application.Features.Products.Commands
+{
+    public class AddProductsCommand : IRequest<ProductsDto>
+    {
+        public AddProductsCommand(string userId, List<ProductCreationDto> productCreation)
+        {
```
Comment (chimaihueze) [inline src/Hng.Application/Features/Products/Commands/AddProductsCommand.cs:14]: This command is for adding Product, it has no business taking userId as a parameter/argument. If you want to assign a user to a product. Do it in the handler.

**Review by chimaihueze — CHANGES_REQUESTED** [review]
Write your unit tests as well.

File: src/Hng.Application/Features/Products/Commands/AddProductsCommand.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Hng.Application.Features.Products.Commands
+{
+    public class AddProductsCommand : IRequest<ProductsDto>
+    {
+        public AddProductsCommand(string userId, List<ProductCreationDto> productCreation)
```
Comment (Godhanded) [inline src/Hng.Application/Features/Products/Commands/AddProductsCommand.cs:13]: don't collect the userId from the request, use the getloggedinuserId function in authentication service class in hng.infrastructure

File: src/Hng.Application/Features/Products/Handlers/AddProductsHandler .cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        }
+        public async Task<ProductsDto> Handle(AddProductsCommand request, CancellationToken cancellationToken)
+        {
+            var productResponse = new ProductsDto();
+
+            foreach (var item in request.productBody)
+            {
+                var product = _mapper.Map<Product>(item);
+                product.Id = Guid.NewGuid();
+                product.UserId = Guid.Parse(request.UserId);
```
Comment (Godhanded) [inline src/Hng.Application/Features/Products/Handlers/AddProductsHandler .cs:28]: use the GetLoggedInUserId function in hng.infrastructure/services/authenticationservice  to get the id of the user

**Review by Godhanded — CHANGES_REQUESTED** [review]

**Comment by AlakaTahir** [comment]
Another branch created

**Pull request closed without merging**